From 781e909bf275df7bc972a647ee0a551453dd1b4b Mon Sep 17 00:00:00 2001
From: progier389 <progier@redhat.com>
Date: Tue, 10 Sep 2024 16:29:00 +0200
Subject: [PATCH] Issue 6321 - lib389 get_db_lib function may returns the wrong
 db type (#6322)

* Issue 6321 - lib389 get_db_lib function may returns the wrong db type

get_db_lib returns the default db type instead of the configured one because of caught exceptions
Fix the import issue if instance is online
Fix the DSEldif parameter for the offline case

Issue: #6321

Reviewed by: @tbordaz (Thanks!)
---
 .../suites/lib389/config_compare_test.py      | 54 +++++++++++++++++++
 src/lib389/lib389/__init__.py                 |  4 +-
 2 files changed, 56 insertions(+), 2 deletions(-)

diff --git a/dirsrvtests/tests/suites/lib389/config_compare_test.py b/dirsrvtests/tests/suites/lib389/config_compare_test.py
index 9ed4da6c8..3eb3d0ea4 100644
--- a/dirsrvtests/tests/suites/lib389/config_compare_test.py
+++ b/dirsrvtests/tests/suites/lib389/config_compare_test.py
@@ -9,11 +9,17 @@
 import os
 import pytest
 
+from contextlib import suppress
 from lib389.topologies import topology_i2
 from lib389.config import Config
+from lib389.dseldif import DSEldif
 
 pytestmark = pytest.mark.tier1
 
+# set_db_type_and_state fixture parameters
+db_types_and_states = [('bdb,stopped'), ('mdb,stopped'), ('bdb,started'), ('mdb,started')]
+
+
 def test_config_compare(topology_i2):
     """
     Compare test between cn=config of two different Directory Server intance.
@@ -42,6 +48,54 @@ def test_config_compare(topology_i2):
     assert Config.compare(st1_config, st2_config)
 
 
+@pytest.fixture(scope="function", params=db_types_and_states)
+def set_db_type_and_state(topology_i2, request):
+    """
+    Stop standalone1 instance and save its dse.ldif then restore things at teardown.
+    """
+    inst = topology_i2.ins.get('standalone1')
+    dbtype,state = request.param.split(',')
+    inst.stop()
+    becfgdn = 'cn=config,cn=ldbm database,cn=plugins,cn=config'
+    becfgattr = 'nsslapd-backend-implement'
+    save_dse_ldif = DSEldif(inst)
+    dse_ldif = DSEldif(inst)
+    dse_ldif.replace(becfgdn, becfgattr, dbtype)
+    if state == 'started':
+        inst.start()
+
+    def fin():
+        save_dse_ldif._update()
+        inst.start()
+
+    request.addfinalizer(fin)
+    return (dbtype,state)
+
+
+def test_get_db_lib(request, topology_i2, set_db_type_and_state):
+    """
+    Check that get_db_lib() returns the configured database type.
+
+    :id: 04205590-6c70-11ef-bfae-083a88554478
+
+    :setup: two isolated directory servers. standalone1 is set with
+            a specified db type and instance state.
+
+    :steps: 1. Clear get_db_lib cache
+            2. Check that inst.get_db_lib() returns the expected db type
+
+    :expectedresults: 1. Success
+                      2. Success
+    """
+
+    inst = topology_i2.ins.get('standalone1')
+    dbtype,state = set_db_type_and_state
+
+    with suppress(AttributeError):
+        del inst._db_lib
+    assert inst.get_db_lib() == dbtype
+
+
 if __name__ == '__main__':
     # Run isolated
     # -s for DEBUG mode
diff --git a/src/lib389/lib389/__init__.py b/src/lib389/lib389/__init__.py
index 1ac9770b0..65e70c1dd 100644
--- a/src/lib389/lib389/__init__.py
+++ b/src/lib389/lib389/__init__.py
@@ -879,11 +879,11 @@ class DirSrv(SimpleLDAPObject, object):
         with suppress(AttributeError):
             return self._db_lib
         with suppress(Exception):
-            from backend import DatabaseConfig
+            from lib389.backend import DatabaseConfig
             self._db_lib = DatabaseConfig(self).get_db_lib()
             return self._db_lib
         with suppress(Exception):
-            dse_ldif = DSEldif(None, self)
+            dse_ldif = DSEldif(self)
             self._db_lib = dse_ldif.get(DN_CONFIG_LDBM, "nsslapd-backend-implement", single=True)
             return self._db_lib
         return get_default_db_lib()
-- 
2.49.0

